Annotate module API functions in redismodule.h for use with -fno-common

natoscott opened the pull request:
In order to keep the redismodule.h self-contained but still usable with
gcc v10 and later, annotate each API function tentative definition with
the __common__ attribute.  This avoids the 'multiple definition' errors
modules will otherwise see for all API functions at link time.

Further details at https://gcc.gnu.org/gcc-10/porting_to.html

agmt commented:
Is it a problem for other compilers? If so, what about using `extern` (for all non-main files which must have some marker like `#define REDISMODULE_ADDITIONAL_FILE`)?

natoscott commented:
> Is it a problem for other compilers? If so, what about using `extern` (for all non-main files which must have some marker like `#define REDISMODULE_ADDITIONAL_FILE`)?

@agmt the situation with this header is more complicated than 'extern' allows us to resolve, unfortunately.

The header is designed to be completely standalone such that the header file actually creates the functions its describing within the modules object files that use those functions.  There is no function someplace else that we can declare an 'extern' reference to as a result, and AIUI it was the intent of the author of this header that this not be needed.

I'm not sure if it affects other compilers - seems like not, so far, none have this new strictness gcc-10 has now introduced by default.  The annotations I added simply end up making the header provide its previous behaviour, for gcc.

yossigo commented:
@natoscott Thanks for reporting this! This issue with `redismodule.h` already came up in the past a few times and at least in some cases was solved in other ways. For example: https://github.com/RediSearch/RediSearch/blob/master/src/redismodule.h

I tend to think the "correct" solution is really to properly define all function pointers as extern, and require the user to explicitly invoke a macro that handles actual definition in a compilation unit of choice. This would have avoided the problem in the first place, but will break compatibility of course.

If we do stick with your suggestion as a simpler and backwards compatible solution for GCC 10.x, I'd at least consider creating some macro for the attribute settings to have better control, disable it on compilers that don't need or worse, don't accept it, etc.

natoscott commented:
@yossigo I agree re the technically correct solution, but tend to think backwards compatibility should win out here now, with so many modules written and since its relatively easily achieved.

I've updated the redisraft PR - if that redismodule.h seems OK, I'll resolve the conflicts and update this PR too.

Related: https://github.com/RedisLabs/redisraft/pull/58

natoscott commented:
> I've updated the redisraft PR - if that redismodule.h seems OK, I'll resolve the conflicts and update this PR too.

I went ahead and merged in all the recent changes to redismodule.h then recreated the original change, along with the conditional macros discussed earlier.

agmt commented:
@natoscott I don't see any backward-compatibility issue if extern will appear only in case if it's explicitly defined. All existing code, which just includes `redismodule.h`, remains to add variables into object file (no `extern`).

yossigo commented:
@natoscott Thanks, this looks much better now!

I think in the long run (maybe for 7.0) we should really fix this properly and provide a way for users to specify where vars are defined. I don't consider backwards compatibility a big issue here because:
1. It only affects compilation - ABI remains intact.
2. At the moment users generally copy `redismodule.h` over so if things do break, they're likely to break due to an explicit upgrade step.

I tend to think of this as a good interim solution.
@oranagra any other thoughts on this matter?

oranagra commented:
this topic was discussed in:
- #3455 suggested to add an extern and a C file
- #6330 added some macros to allow adding extern in one compilation unit and not in others.
- #6617 added more macros to add `__attribute__ ((weak))`

I voiced my opinion in the last one, if you want to see what i dislike.
And the suggestion i gave there is somewhat similar to this PR is headed.
So i'd like to suggest a few changes.

my concerns:
- I don't want to change these lines again and again (i care about the blame log and merge conflicts), so if we're gonna change all of them, let's have them ready for whatever comes in the future, i.e. i'd like to add another define that's used before the function declaration, so that we can use it to add `extern`.
- I rather the C file that includes this header be the one that sets the decorators (be it `extern`, `__attribute__` or whatever), so even if we do want to add some attributes out of the box, we should have them defined only if not already defined, so that if the file that did the include wants to use something else, we let it have its way.
- maybe this is the opportunity to get rid of the `REDISMODULE_API_FUNC` "function" macro. personally the parenthesis are giving my brain parsing error. reading that line, it's not a C syntax. my brain does seem to be able to skip simple defines like the `REDISMODULE_ATTR_COMMON` that's added here.

So what i would suggest is either this:
```c
#if !defined REDIS_MODULE_PRE_API and defined REDISMODULE_API_DECLARE
#define REDIS_MODULE_PRE_API extern
#endif

REDIS_MODULE_PRE_API void *(*RedisModule_Alloc)(size_t bytes) REDISMODULE_POST_API;
```
So that one C file includes the header without any define, and then it creates the function pointers (this is backwards compatible with projects that had only one C file interacting with redis), And other C files that just need forward declarations, define `REDISMODULE_API_DECLARE` before including the redismodule.h.

or if we wanna go with the common or weak attributes, do this:
```c
#if !defined REDISMODULE_POST_API
#define REDISMODULE_POST_API __attribute__((__common__))
#endif

REDIS_MODULE_PRE_API void *(*RedisModule_Alloc)(size_t bytes) REDISMODULE_POST_API;
```

either way, the project that includes the header can revert to other ways since it has full control on the decorators.

p.s. obviously add a similar trick for REDISMODULE_ATTR_UNUSED and REDISMODULE_ATTR_PRINTF.

natoscott commented:
@yossigo @oranagra 

> this topic was discussed in:

Hmm, what a problem this whole area has become. :(

> So i'd like to suggest a few changes.

I think these can be catered for relatively simply - see the suggestion below.

>     * I don't want to change these lines again and again (i care about the blame log and merge conflicts), so if we're gonna change all of them, let's have them ready for whatever comes in the future, i.e. i'd like to add another define that's used before the function declaration, so that we can use it to add `extern`.

Agreed.

>     * I rather the C file that includes this header be the one that sets the decorators (be it `extern`, `__attribute__` or whatever), so even if we do want to add some attributes out of the box, we should have them defined only if not already defined, so that if the file that did the include wants to use something else, we let it have its way.

Agreed.

>     * maybe this is the opportunity to get rid of the `REDISMODULE_API_FUNC` "function" macro. 

Strictly speaking, removing it breaks backwards compatibility (someone, somewhere will have used this macro).  I've kept it for now.

> ```c
> #if !defined REDIS_MODULE_PRE_API and defined REDISMODULE_API_DECLARE
> #define REDIS_MODULE_PRE_API extern
> #endif

Here's another way with less line noise IMO, but keeping 100% backward compatibility.  REDISMODULE_API can be defined outside of the header to add the 'extern', and REDISMODULE_ATTR_COMMON can be defined to nothing earlier too.

```--- a/redismodule.h
+++ b/redismodule.h
@@ -383,11 +383,15 @@ typedef long long mstime_t;
 /* Macro definitions specific to individual compilers */
 #ifdef __GNUC__
 #define REDISMODULE_ATTR_UNUSED __attribute__((unused))
+#ifndef REDISMODULE_ATTR_COMMON
 #define REDISMODULE_ATTR_COMMON __attribute__((__common__))
+#endif
 #define REDISMODULE_ATTR_PRINTF(idx,cnt) __attribute__((format(printf,idx,cnt)))
 #else
 #define REDISMODULE_ATTR_UNUSED
+#ifndef REDISMODULE_ATTR_COMMON
 #define REDISMODULE_ATTR_COMMON
+#endif
 #define REDISMODULE_ATTR_PRINTF(idx,cnt)
 #endif
 
@@ -448,13 +452,17 @@ typedef struct RedisModuleTypeMethods {
     RedisModule_GetApi("RedisModule_" #name, ((void **)&RedisModule_ ## name))
 
 #define REDISMODULE_API_FUNC(x) (*x)
+#ifndef REDISMODULE_API
+#define REDISMODULE_API(type,x) type REDISMODULE_API_FUNC(x)
+#endif
 
+REDISMODULE_API(void *, RedisModule_Alloc)(size_t bytes) REDISMODULE_ATTR_COMMON;
+REDISMODULE_API(void *, RedisModule_Realloc)(void *ptr, size_t bytes) REDISMODULE_ATTR_COMMON;
+REDISMODULE_API(void, RedisModule_Free)(void *ptr) REDISMODULE_ATTR_COMMON;
+REDISMODULE_API(void *, RedisModule_Calloc)(size_t nmemb, size_t size) REDISMODULE_ATTR_COMMON;
+REDISMODULE_API(char *, RedisModule_Strdup)(const char *str) REDISMODULE_ATTR_COMMON;

/* [...snip other similar APIs, but here's one with multiple attributes...] */

+REDISMODULE_API(void, RedisModule_Log)(RedisModuleCtx *ctx, const char *level, const char *fmt, ...) REDISMODULE_ATTR_PRINTF(3,4) REDISMODULE_ATTR_COMMON;
```

> p.s. obviously add a similar trick for REDISMODULE_ATTR_UNUSED and REDISMODULE_ATTR_PRINTF.

I'm not quite sure what you have in mind there (not obvious to me) - see the RedisModule_Log API example above.  I tend to think we must have separate macros for each trailing attribute - some of which could be defined away, but it's also more readable.

If this approach is generally acceptable, I'll make the change throughout the header and update the PR.

oranagra commented:
@natoscott i personally really dislike these function style macros which take return type and name. IMHO it makes the reading of these lines more difficult. the C syntax for function pointers is complicated enough, so i rather stick to it and not add any new parenthesis and macros that generate the whole thing.
Personally i think that adding a PRE and POST defines doesn't diverge these lines from C syntax and more readable to me.

The reason i insist on PRE and POST and not COMMON or EXTERN defines is because the user can add any decorators he wants (i.e. `weak` attribute instead of `common`).

Regarding the UNUSED and PRINTF defines, what i meant is that they should be defined only if not already defined. so that we don't force them on the user, and he has a way to disable or modify them from the C file.

regarding the REDISMODULE_API_FUNC macro, i really don't see what it can be used for. if anyone has any idea, please enlighten me.
@itamarhaber can you please run a quick grep to see if any of the modules you're familiar with messes with this define?
p.s. like Yossi said, even if we break one module, i don't think that's that bad (unless the removal of this macro induces some limitation that cannot be solved without it).

oranagra commented:
@yossigo please share your opinion on my proposal.
@itamarhaber can you check about any module explicitly messing with REDISMODULE_API_FUNC?

itamarhaber commented:
@oran grep and github search do not seem to return anything to worry about.

natoscott commented:
I've updated the code with before/after macros, allowing them to be set outside the header, and defaulting to values that maintain the previous behaviour.  As requested, the REDISMODULE_API_FUNC is removed despite the backwards compatibility issue this introduces.

oranagra commented:
thanks @natoscott LGTM.
just to be on the safe side, can you please test to see how this behaves in a dummy project composed of multiple source files that interact with the API (both C and C++, with new and old GCC).

The C++ files would have to do `extern "C"` and that's their burden IMHO. but i'm wondering if we should make any other adjustments to make it work better for C projects out of the box, and also make sure that whatever needs to be adjusted can be done from outside the header file, or at the very least without modifying every API line.

itamarhaber commented:
@sewenew would you mind, if you have the time, to take a look and see whether your CPP modules work as expected with this change.

sewenew commented:
@itamarhaber This seems not work for C++. It seems `__common__` is a C specific attribute. I tried to compile my C++ module with this new redismodule.h file. It failed to compile and report the following errors:

```
redis-protobuf/src/sw/redis-protobuf/redismodule.h:460:59: error: '__common__' attribute is not supported in C++
REDISMODULE_API void * (*RedisModule_Alloc)(size_t bytes) REDISMODULE_ATTR;
                                                          ^
redis-protobuf/src/sw/redis-protobuf/redismodule.h:457:26: note: expanded from macro 'REDISMODULE_ATTR'
#define REDISMODULE_ATTR REDISMODULE_ATTR_COMMON
                         ^
redis-protobuf/src/sw/redis-protobuf/redismodule.h:386:48: note: expanded from macro 'REDISMODULE_ATTR_COMMON'
#define REDISMODULE_ATTR_COMMON __attribute__((__common__))
                                               ^
```

My compiler is `Apple clang version 11.0.3 (clang-1103.0.32.62)`. I don't have a GCC 10 compiler so far, and cannot test with it.

Regards

sewenew commented:
In fact, the old redismodule.h does not work with C++ either. C++ strictly distinguishes definition and declaration, and the compiler always complains about multiple definition errors.

In order to solve the problem, I split redismodule.h into a .h file and .cpp file. Declare these function pointers as extern in header file, and define them in the source file. You can check [redismodule.h](https://github.com/sewenew/redis-protobuf/blob/master/src/sw/redis-protobuf/redismodule.h) and [redismodule.cpp](https://github.com/sewenew/redis-protobuf/blob/master/src/sw/redis-protobuf/redismodule.cpp) for detail.

Regards

oranagra commented:
@sewenew the point is that with the new header you can do what you did without editing the header file.

in one C file you do:
```c
#define REDISMODULE_API
#define REDISMODULE_ATTR
#include "redismodule.h
```

and in the others you do:
```c
#define REDISMODULE_API extern
#define REDISMODULE_ATTR
#include "redismodule.h"
```

i'm hoping that for C projects that use multiple files the default header would be enough with no need to define anything.

sewenew commented:
@oranagra Thanks for the explanation!

I followed your instruction, and it works with two minor fixes:

1. There's a redefinition of `RedisModule_Alloc` on [line 461](https://github.com/natoscott/redis/blob/fno-common/src/redismodule.h#L461) of the new redismodule.h file, and need to be removed.

2. `REDISMODULE_ATTR_PRINTF` doesn't work with [C++ variadic template](https://en.wikipedia.org/wiki/Variadic_template), since there's no format string as parameter in this case (check [this](https://stackoverflow.com/questions/12573968/how-to-use-gccs-printf-format-attribute-with-c11-variadic-templates) for reference). I use the following code to fix the problem. FYI.

```
 /* Macro definitions specific to individual compilers */
 #ifdef __GNUC__
 #define REDISMODULE_ATTR_UNUSED __attribute__((unused))
 #define REDISMODULE_ATTR_COMMON __attribute__((__common__))
+#ifndef REDISMODULE_ATTR_PRINTF
 #define REDISMODULE_ATTR_PRINTF(idx,cnt) __attribute__((format(printf,idx,cnt)))
+#endif
 #else
 #define REDISMODULE_ATTR_UNUSED
 #define REDISMODULE_ATTR_COMMON
 #define REDISMODULE_ATTR_PRINTF(idx,cnt)
 #endif
```

So that I can use something like the following to disable it:

```
 #define REDISMODULE_API
 #define REDISMODULE_ATTR
+#define REDISMODULE_ATTR_PRINTF(idx,cnt)
 #include "redismodule.h"
```

Regards

oranagra commented on src/redismodule.h:
```diff
@@ -380,6 +380,17 @@ typedef struct RedisModuleLoadingProgressInfo {
 
 typedef long long mstime_t;
 
+/* Macro definitions specific to individual compilers */
+#ifdef __GNUC__
+#define REDISMODULE_ATTR_UNUSED __attribute__((unused))
+#define REDISMODULE_ATTR_COMMON __attribute__((__common__))
+#define REDISMODULE_ATTR_PRINTF(idx,cnt) __attribute__((format(printf,idx,cnt)))
+#else
+#define REDISMODULE_ATTR_UNUSED
+#define REDISMODULE_ATTR_COMMON
+#define REDISMODULE_ATTR_PRINTF(idx,cnt)
+#endif
```
```suggestion
/* Macro definitions specific to individual compilers */
#ifdef __GNUC__
#  ifndef REDISMODULE_ATTR_UNUSED
#    define REDISMODULE_ATTR_UNUSED __attribute__((unused))
#  endif
#  ifndef REDISMODULE_ATTR_COMMON
#    define REDISMODULE_ATTR_COMMON __attribute__((__common__))
#  endif
#  ifndef REDISMODULE_ATTR_PRINTF
#    define REDISMODULE_ATTR_PRINTF(idx,cnt) __attribute__((format(printf,idx,cnt)))
#  endif
#else
#  ifndef REDISMODULE_ATTR_UNUSED
#    define REDISMODULE_ATTR_UNUSED
#  endif
#  ifndef REDISMODULE_ATTR_COMMON
#    define REDISMODULE_ATTR_COMMON
#  endif
#  ifndef REDISMODULE_ATTR_PRINTF
#    define REDISMODULE_ATTR_PRINTF(idx,cnt)
#  endif
#endif
```

natoscott commented:
> @oranagra Thanks for the explanation!
> 
> I followed your instruction, and it works with two minor fixes:
> 
>     1. There's a redefinition of `RedisModule_Alloc` on [line 461](https://github.com/natoscott/redis/blob/fno-common/src/redismodule.h#L461) of the new redismodule.h file, and need to be removed.

Thanks - fixed.

>     2. `REDISMODULE_ATTR_PRINTF` doesn't work with [C++ variadic template](https://en.wikipedia.org/wiki/Variadic_template), since there's no format string as parameter in this case (check [this](https://stackoverflow.com/questions/12573968/how-to-use-gccs-printf-format-attribute-with-c11-variadic-templates) for reference). I use the following code to fix the problem. FYI.
> 

This doesn't sound right to me - AIUI the stackoverflow link is not the same thing as what we're doing here.  In that post the OP describes attempts to annotate the C++ code to match the C code its calling, its not about annotation on the underlying C code (which is what we're doing here). 

It's important that the RedisModule_Log function is passed a constant string for the formatter - not just to be able to check the parameter types but also to ensure we are never in the situation that user/input data might be used as the formatting string, as that can cause crashes and/or security problems.

Can you post the warnings/errors you see, and explain a bit more how your C++ API / wrapper code ends up calling RedisModule_Log?  (maybe with some sample code?)  If the C++ code is producing the final formatted string (?) you might be able to just pass "%s" as the first parameter.


This is all quite off-topic for the original issue being solved here - making the header C++ friendly should be a separate PR.

> So that I can use something like the following to disable it:

IMO the suggestions to 'solve' this are increasingly ugly and allow possibly dangerous ways of calling the RedisModule_Log routine that would otherwise have been perfectly well protected.  The case for such changes should be argued in a separate PR and not piggy-backed onto this unrelated issue.

Thanks.

sewenew commented:
@natoscott  Yes, I can pass a format string, e.g. %s, to work around it. Not a big deal.

Regards

yossigo commented:
Looks good to me now, I've tested it on gcc 4.x, 9.x, 10.x (C and C++) and clang 10.x. It seems like clang (like gcc <= 9.x) doesn't need any special attribute for C and doesn't support `common` for C++, so I've added a check for that.

oranagra commented:
@natoscott thank you for suggesting it, and thank you for keeping up with my requests.
I hope we'll have a smoother sailing from here on. this header file was causing pain for many.
